docs/design-documents/promtail-scraping-objectstore.md

```diff
@@ -0,0 +1,89 @@
+
+Cortex already has the [implementation](https://github.com/cortexproject/cortex/blob/master/pkg/chunk/aws/s3_storage_client.go) for Objecstore including S3 and GCS. We can try to re-use most of it. Still, we have to develop code for `file watching` and updating the `positions.yaml` file.  Following are the few challenges which we should try to address.
+
+### There is no appending of lines to file in object store
+
+In an object store there is no appending of lines to a file in bucket. Any file modified will be newly 
+uploaded to the bucket with the same name. So, we would have to read the entire file again from the begining since we will not be aware if only new lines were appended to the modified file or the file is completely having new content
+
+### Where should we keep positions.yaml file ?
+
+Since, the files we scrape is remotely placed in a objecstore, should we keep the `positions.yaml` also in the same place? Advantage of keeping the `positions.yaml` within the objecstore will help us to recover from the state where the scraping was left when promtail is running in docker container and the container crashes. However, `user` reading the files from the Object store would also need write access to the Objecstore. Getting write access may not be possible in some cases where people doesn't want to risk modifying the files by giving write access.
+
+### How should we support reading of compressed files ?
+
+If the files in the Objectstore is stored in a compressed format, promtail should have support for reading the compressed data and we should make sure we support most of the compression algorithms like we already do for chunks(Snappy, LZ4, Gzip) 
+
+### How we should identify and watch newly added files ?
+
+To identify and to start tailing the newly added files, we should list the files in the Object store at some desired intervals. How frequently we should this? Once we add the file to `watch`, to identify any changes to the file we should frequently check the modified `datetime` of the file. The list of files might be large and each watch will be running in a `go routine` and each `go routine` making frequent `http` requests to the Obejctstore. We should make sure this doesn't affect promtail's performance.
+
+## Proposed Solution for the above mentioned points
+
+> There is no appending of lines to file in object store
+
+In the positions.yaml along with the `offset` we should also store file modified `datetime`. If we see that date is modified we should reset the offset and tail the contents from the begining. This is the same behaviour in local filesystem as well. If a file is `opened`, `hpcloud tail` will tail the file from the begining.
```

Comment:
What's the problem if we resume where the offset was when a file is modified ? I think it's fine to expect a file to be updated with a bigger content.

Reply:
There are two scenarios here. One, file is just appended with more content and uploaded. Two, full content of the file is modified. How will we identify this? If the whole file is modified then we are expected to tail from the beginning

Reply:
Can you do some research to see how it work for AWS loadbalancer logs which is the essence of this feature. I wondering if they are just dropping file.

I'm ok to no support update for the first version too and leave that for later. I don;'t think file update is a big requirement.

/cc @Punkoivan

Reply:
>If we see that date is modified we should reset the offset and tail the contents from the begining

I was thinking that modified time will be updated if some content was just appended.

We can store timestamp for the last line in the file and the compare to appended/replaced line's timestamp.
Pseudo code:
```
if previous_line_ts > new_line_ts
then content was fully replaced
if previous_line_ts < new_line_ts
then content was added.
```
That's based on the statement "Loki doesn't support out of order message". In other words even if we have new lines with older timestamp, it will be declined.

I assume it's bad approach which will lead to a huge overhead, especially in case where we have hundreds of files, but might be helpful to develop good approach :)